middleware/decrypt.py

```diff
@@ -0,0 +1,90 @@
+"""Identify and decrypt Crypt4GH keys and files.
+
+Decrypts crypt4GH files given a list of files and places the output in a specified
+directory. If a file is not Crypt4GH-encrypted, it moves the file to the output directory
+without alteration. If a Crypt4GH file is included, the private key associated with that
+file must be provided.
+
+Example:
+    python3 decrypt.py --output-dir /outputs/ file.txt file.c4gh sk.sec pk.pub
+"""
+import os
+from argparse import ArgumentParser
+from pathlib import Path
+from crypt4gh.lib import decrypt  # type: ignore
+from crypt4gh.keys import get_private_key  # type: ignore
+
+
+def get_private_keys(file_paths: list[Path]) -> list[bytes]:
+    """Retrieve private keys from a list of files.
+
+    Args:
+        file_paths (list[Path]): A list of file paths.
+
+    Returns:
+        list[bytes]: A list of retrieved private keys as byte objects.
+    """
+    private_keys = []
+    for file_path in file_paths:
+        try:
+            key = get_private_key(Path(file_path), lambda x: '')
+            private_keys.append(key)
+        except ValueError:
+            continue
+    return private_keys
+
+
+def decrypt_files(
+        file_paths: list[Path],
+        private_keys: list[bytes],
+        output_path: Path) -> None:
+    """Decrypt files and save to specified output directory.
+
+    Args:
+        file_paths (list[Path]): A list of file paths.
+        private_keys (list[bytes]): A list of private keys as byte objects.
+        output_path: (Path): Directory to place decrypted files in.
+
+    Raises:
+        ValueError: If no private key for a Crypt4GH file is provided.
+    """
+    for file_path in file_paths:
+        with open(file_path, "rb") as f_in:
+            filename = file_path.name
+            try:
+                with open(output_path/filename, "wb") as f_out:
+                    decrypt([(0, pk, "") for pk in private_keys], f_in, f_out)
+            except ValueError as e:
+                if str(e) == "Not a CRYPT4GH formatted file":
```

Comment:
@jvkersch Currently encrypted and unencrypted files are distinguished through attempting to decrypt every file with the crypt4gh Python library. The decrypt function checks to see if the magic is at the beginning of the file ([here](https://github.com/EGA-archive/crypt4gh/blob/master/crypt4gh/header.py#L43)). If the magic is not present, it throws the message "Not a CRYPT4GH formatted file" (so I continue iterating through the files). I think the fact that I just have a continue statement makes this unclear. Should I just check for the magic myself? I think that could be clearer.

Reply:
Opted to add an inline comment rather than checking for the magic myself because the crypt4gh checks for the magic and I wanted to avoid opening the file twice.